### ISSUE TITLE
Native slicing overwrite is inconsistent with deprecated overwrite() op, both are inconsistent with cpu reference

### ISSUE THREAD

**Issue by arielge** [torch-spyre/torch-spyre#2561]
### What happened?

Spyre outputs are inconsistent with cpu reference when overwriting values in model KV cache:

1. For the _deprecated_ (#2488) `torch.ops.spyre.overwrite()`, the inconsistencies emerge after exceeding the torch dynamo `accumulated_recompile_limit` of 256. This op is currently used in the [hf_adapters](https://github.com/torch-spyre/hf-adapters) repo.

2. When using native torch slicing for the same operation, outputs are inconsistent with cpu across KV cache offset values.

Max diff vs. cpu:
```
 overwrite:  under(<256) = 0.0117   over(>=256) = 3.8174
 slice:  under(<256) = 2.7686   over(>=256) = 2.3398
```

### What did you expect to happen?

Native slicing should replace the deprecated overwrite() with similar behavior, both should be numerically consistent with CPU reference outputs.

### How can we reproduce it?

```python
import torch

BATCH, NUM_KV_HEADS, HEAD_DIM, HIDDEN = 1, 8, 128, 1024
DTYPE, SEQ = torch.float16, 64
ACCUM = torch._dynamo.config.accumulated_recompile_limit  # 256
N = ACCUM + 16


class FusedFillWrite(torch.nn.Module):
    """RMSNorm -> k_proj -> write a width-1 K slice into cache at `offset`."""
    def __init__(self, strategy):
        super().__init__()
        self.strategy = strategy
        self.ln_w = torch.nn.Parameter(torch.randn(HIDDEN, dtype=DTYPE))
        self.k_proj = torch.nn.Linear(HIDDEN, NUM_KV_HEADS * HEAD_DIM, bias=False).to(DTYPE)

    def forward(self, x, cache, token_index, offset):
        var = (x * x).mean(-1, keepdim=True)
        eps = torch.full((1,), 1e-6, dtype=x.dtype, device=x.device)
        h = self.ln_w * (x * torch.rsqrt(var + eps))
        bsz, sl, _ = x.shape
        k = self.k_proj(h).view(bsz, sl, NUM_KV_HEADS, HEAD_DIM).transpose(1, 2)
        k_slice = k[:, :, token_index:token_index + 1, :]
        if self.strategy == "overwrite":
            torch.ops.spyre.overwrite(input=k_slice, output=cache, dims=[2], offsets=[offset])
        else:  # "slice"
            cache[:, :, offset:offset + 1, :] = k_slice
        return cache


def make_x(i):
    g = torch.Generator().manual_seed(1000 + i)
    return torch.randn(BATCH, SEQ, HIDDEN, generator=g).to(DTYPE)


def run(strategy):
    torch.manual_seed(0)
    ref = FusedFillWrite("slice").eval().requires_grad_(False)

    # CPU reference (plain eager).
    cpu = torch.zeros(BATCH, NUM_KV_HEADS, N, HEAD_DIM, dtype=DTYPE)
    for i in range(N):
        ref(make_x(i), cpu, i % SEQ, i)
    cpu = cpu.float()

    # Spyre, one compiled frame, all N offsets, same weights as the reference.
    torch._dynamo.reset()
    mod = FusedFillWrite(strategy).eval().requires_grad_(False)
    mod.load_state_dict(ref.state_dict())
    mod = mod.to("spyre")
    cache = torch.zeros(BATCH, NUM_KV_HEADS, N, HEAD_DIM, dtype=DTYPE, device="spyre")
    compiled = torch.compile(mod, dynamic=False)
    for i in range(N):
        compiled(make_x(i).to("spyre"), cache, i % SEQ, i)
    sp = cache.to("cpu").float()

    under = over = 0.0
    for i in range(N):
        d = (sp[:, :, i, :] - cpu[:, :, i, :]).abs().max().item()
        if i >= ACCUM:
            over = max(over, d)
        else:
            under = max(under, d)
    return under, over


for strat in ("overwrite", "slice"):
    under, over = run(strat)
    print(f"{strat:>10}:  under(<{ACCUM}) = {under:.4f}   over(>={ACCUM}) = {over:.4f}")
```

### Any environmental details we need to know?

torch-spyre commit `8028f91`

### Relevant log output

```Shell
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256] torch._dynamo hit config.accumulated_recompile_limit (256)
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]    function: 'forward' (/home/arielgera/hf_adapters/repro_v3_temp.py:19)
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]    last reason: 0/255: offset == 255                                            # torch.ops.spyre.overwrite(input=k_slice, output=cache, dims=[2], offsets=[offset])  # hf_adapters/repro_v3_temp.py:27 in forward
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256] User stack trace:
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]   File "/home/arielgera/hf_adapters/repro_v3_temp.py", line 27, in forward
W0608 15:17:21.138000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]     torch.ops.spyre.overwrite(input=k_slice, output=cache, dims=[2], offsets=[offset])

W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256] torch._dynamo hit config.accumulated_recompile_limit (256)
W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]    function: 'forward' (/home/arielgera/hf_adapters/repro_v3_temp.py:19)
W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]    last reason: 0/255: offset == 255                                            # cache[:, :, offset:offset + 1, :] = k_slice  # hf_adapters/repro_v3_temp.py:29 in forward
W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256] User stack trace:
W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]   File "/home/arielgera/hf_adapters/repro_v3_temp.py", line 29, in forward
W0608 15:25:45.326000 11294 torch/_dynamo/convert_frame.py:1743] [0/256]     cache[:, :, offset:offset + 1, :] = k_slice
```

**Comment by flaviabeo** [comment]
I was able to reproduce the error:
```bash
  warn(depr_msg, FutureWarning, stacklevel=1)
/home/flaviabeodebayser/dt-inductor/torch-spyre/torch_spyre/_inductor/lowering.py:749: FutureWarning: torch.ops.spyre.overwrite is deprecated. Use standard PyTorch operations like output[indices] = input or output[indices].copy_(input). Please report any incompatibilities.
  warn(depr_msg, FutureWarning, stacklevel=1)
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256] torch._dynamo hit config.accumulated_recompile_limit (256)
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256]    function: 'forward' (/home/flaviabeodebayser/dt-inductor/torch-spyre/test_2561.py:17)
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256]    last reason: 0/255: offset == 255                                            # torch.ops.spyre.overwrite(input=k_slice, output=cache, dims=[2], offsets=[offset])  # torch-spyre/test_2561.py:25 in forward
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256] User stack trace:
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256]   File "/home/flaviabeodebayser/dt-inductor/torch-spyre/test_2561.py", line 25, in forward
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256]     torch.ops.spyre.overwrite(input=k_slice, output=cache, dims=[2], offsets=[offset])
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256] To log all recompilation reasons, use TORCH_LOGS="recompiles".
W0608 18:31:59.977000 1103811 torch/_dynamo/convert_frame.py:1743] [0/256] To diagnose recompilation issues, see https://docs.pytorch.org/docs/main/user_guide/torch_compiler/compile/programming_model.recompilation.html
/home/flaviabeodebayser/dt-inductor/torch-spyre/torch_spyre/_inductor/lowering.py:749: FutureWarning: torch.ops.spyre.overwrite is deprecated. Use standard PyTorch operations like output[indices] = input or output[indices].copy_(input). Please report any incompatibilities.
  warn(depr_msg, FutureWarning, stacklevel=1)
```
So will be looking into this - The branches that contain fixes for offsets still do not fix this one.

**Comment by yoheiueda** [comment]
@arielge Thank you for reporting this issue! It seems the existing overwrite op already has some limitations. Since symbolic shapes/offsets are not supported, different offsets always trigger recompilation. That may also indicate there is some issue when offsets >= 256.

@flaviabeo Thank you for working on this. I’d be happy to help diagnose the issue.

Since this is a numerical correctness issue rather than a compilation error, debugging may be a bit more challenging.

One thing that came to mind is the following transpose-related issue, although I’m not yet sure whether it is directly related:

- #1756

**Comment by flaviabeo** [comment]
> [@arielge](https://github.com/arielge) Thank you for reporting this issue! It seems the existing overwrite op already has some limitations. Since symbolic shapes/offsets are not supported, different offsets always trigger recompilation. That may also indicate there is some issue when offsets >= 256.
> 
> [@flaviabeo](https://github.com/flaviabeo) Thank you for working on this. I’d be happy to help diagnose the issue.
> 
> Since this is a numerical correctness issue rather than a compilation error, debugging may be a bit more challenging.
> 
> One thing that came to mind is the following transpose-related issue, although I’m not yet sure whether it is directly related:
> 
> * [Add padding to non-stick dimension in Restickify to ensure stick-size alignment #1756](https://github.com/torch-spyre/torch-spyre/issues/1756)

Hi @yoheiueda ! sure, will be happy to pair with you on this one.

**Comment by raghukiran1224** [comment]
## Root cause: this is issue #220 (no symbolic offsets in SDSC), surfacing through the recompile limit

I reproduced both failures on hardware (`rganti-spyre-main`, commit `9be4e3d`) and traced the mechanism. Sharing the analysis since it turned out to be a single, well-defined cause — and **not** where the surface symptoms first suggested.

### What the failure pattern actually is

Running the repro with per-offset diffs (N=272), the bad offsets are:

| strategy | bad offsets | matches issue |
|---|---|---|
| `overwrite` | `257..271` | 0.0117 / 3.8174 ✓ |
| `slice` | `0`, `256..271` | 2.7686 / 2.3398 ✓ |

**Offsets `1..255` are correct for both paths** (~0.006). So "slice is wrong across all offsets" isn't quite right — it's `0` plus the `>=256` tail. And the `overwrite` vs `slice` boundary difference (257 vs 256) is just a compile-count difference in when each path crosses the limit.

The single common factor: **the failure starts exactly when `torch._dynamo.config.accumulated_recompile_limit` (256) is hit.** Up to the limit, dynamo specializes each call and the offset is baked as a *concrete constant*. Once the limit is exhausted, dynamo stops specializing and the offset reaches lowering as a **symbol**.

### The exact drop point

I instrumented `superdsc.py` and lowered the recompile limit to 8 to trigger the symbolic path quickly (offsets 0–7 clean, 8+ bad — the boundary tracks the limit). For the cache write past the limit:

```
dim_coord = z0 + 7      as_coeff_Add() = (7, z0)
device_coordinates = [z0 + 7, c0, floor(c1/64), Mod(c1, 64)]
```

where `z0` is the symbolic runtime offset. At `superdsc.py:375`:

```python
dim_offset = int(dim_coord.as_coeff_Add()[0])   # (7, z0)[0] = 7 — z0 is silently dropped
```

`int(dim_coord.as_coeff_Add()[0])` keeps only the **constant coefficient** and discards the symbolic term, so the write address loses the runtime offset. (The `slice` "offset 0" failure is a side effect: the mis-targeted `>=256` writes corrupt cache row 0.)

Worth noting: `compute_coordinates` in `views.py` is **not** at fault — it correctly produces the symbolic coord `z0 + 7`. The drop is solely at the SDSC codegen boundary.

### Why this isn't a one-line fix

That `int(...)` is correct under the current architecture. The whole SDSC codegen boundary concretizes to integers — `core_idx_to_slice_offset` (`sum(arg.offsets.values())`), `offset_as_symbol(addr, ...)` (where `addr` is an `int`), and the DeepTools backend all require concrete per-core addresses. The existing `affine.apply` / tiled-symbol machinery handles *loop-tiled* addresses (`loop_var * tile_stride`), not an arbitrary runtime scalar argument like a KV-cache offset.

This is exactly what the in-tree TODOs call out:

> `superdsc.py` `concretize_expr`: *"final concretization point in the pipeline... TODO(issue#220): once SDSC generation emits `symbolDefinitions_` and `symbolicDimInfo_` for the DeepTools VariableDefinition DAG, this function can be replaced with symbolic expression serialisation and iteration-space sizes can remain symbolic all the way through."*


So **#2561 is a manifestation of #220**: a proper fix needs end-to-end symbolic-offset support (symbol definitions in the SDSC JSON, plumbing through `compute_ops`/`bundle`, and DeepTools backend support) — cross-repo, not a torch-spyre-only patch. The `specialize_int=True` guard for `overwrite` (#2084) works only *until* the recompile limit is exhausted.

### Practical mitigation in the meantime

Since the bug only appears once specialization is exhausted, raising `accumulated_recompile_limit` above the realistic KV-cache length keeps every offset concrete and correct (`cache_size_limit` is already bumped to 1024 in `torch_spyre/__init__.py`). That unblocks hf-adapters for bounded sequence lengths without depending on #220.

### Fast repro for whoever picks this up

`accumulated_recompile_limit = 8` reproduces the symbolic path at offset 8 in ~2 min instead of ~15, which makes iterating on a #220 fix much cheaper.

**Comment by arielge** [comment]
Thanks @raghukiran1224, so this also indicates that in any case there is no reason to keep using the deprecated overwrite(), correct?

**Comment by ani300** [comment]
Yeah, can you update the hf_adapters to stop using it?
